### PR TITLE
Adding Constructors for Convolution and Adaptive mean pooling

### PR DESCRIPTION
**Pull request by gauravghati**
This PR is in reference to issue [#2625](https://github.com/mlpack/mlpack/issues/2625)
Content in PR: 
- Adding Copy and move constructors for Convolution Layer with tests
- Adding Copy and move constructors for Adaptive mean pooling Layer with tests
- Adding Copy and move constructors for Adaptive max pooling Layer with tests
- initialization of outputWidth, outputHeight, reset variables in Adaptive mean and max pooling Layer constructors

### CODE REVIEW

File: src/mlpack/tests/ann_layer_test.cpp [hunk 1]
```diff
@@ -4326,6 +4377,38 @@ TEST_CASE("ConvolutionLayerTestCase", "[ANNLayerTest]")
...
+  // number of images are 3 and number of feature maps are 2.
+  input = arma::mat(8, 3);
+  input << 1 << 446 << 42 << arma::endr
+      << 2 << 16 << 63 << arma::endr
+      << 3 << 13 << 63 << arma::endr
+      << 4 << 21 << 21 << arma::endr
+      << 1 << 13 << 11 << arma::endr
+      << 32 << 45 << 42 << arma::endr
+      << 22 << 16 << 63 << arma::endr
+      << 32 << 13 << 42 << arma::endr;
```
Comment (NippunSharma) [inline src/mlpack/tests/ann_layer_test.cpp:4394]: Use braced initializer list like used here #2708

Comment (gauravghati) [reply src/mlpack/tests/ann_layer_test.cpp]: Thanks for reviewing @NippunSharma, done changes!

**Review by zoq — COMMENTED** [review]
Looks good, can you handle the first issue from the static code analysis report - "outputWidth, outputHeight, reset."?

**Comment by gauravghati** [comment]
Apologies for the delayed response.
I did't get any error or warning for "outputWidth, outputHeight, reset." variables. 
The only errors which I got in `convolution_impl.hpp` during analysis were: 
```
mlpack/src/mlpack/methods/ann/layer/convolution_impl.hpp	218	err	V603 The object was created but it is not being used. If you wish to call constructor, 'this->Convolution::Convolution(....)' should be used.
mlpack/src/mlpack/methods/ann/layer/convolution_impl.hpp	224	err	V603 The object was created but it is not being used. If you wish to call constructor, 'this->Convolution::Convolution(....)' should be used.
mlpack/src/mlpack/methods/ann/layer/convolution_impl.hpp	278	err	V603 The object was created but it is not being used. If you wish to call constructor, 'this->Convolution::Convolution(....)' should be used.
mlpack/src/mlpack/methods/ann/layer/convolution_impl.hpp	347	err	V603 The object was created but it is not being used. If you wish to call constructor, 'this->Convolution::Convolution(....)' should be used.
```

**Comment by zoq** [comment]
> Apologies for the delayed response.
> I did't get any error or warning for "outputWidth, outputHeight, reset." variables.
> The only errors which I got in `convolution_impl.hpp` during analysis were:
> 
> ```
> mlpack/src/mlpack/methods/ann/layer/convolution_impl.hpp	218	err	V603 The object was created but it is not being used. If you wish to call constructor, 'this->Convolution::Convolution(....)' should be used.
> mlpack/src/mlpack/methods/ann/layer/convolution_impl.hpp	224	err	V603 The object was created but it is not being used. If you wish to call constructor, 'this->Convolution::Convolution(....)' should be used.
> mlpack/src/mlpack/methods/ann/layer/convolution_impl.hpp	278	err	V603 The object was created but it is not being used. If you wish to call constructor, 'this->Convolution::Convolution(....)' should be used.
> mlpack/src/mlpack/methods/ann/layer/convolution_impl.hpp	347	err	V603 The object was created but it is not being used. If you wish to call constructor, 'this->Convolution::Convolution(....)' should be used.
> ```

You can click on the static code analysis checks below to see the report. "Not all members of a class are initialized inside the constructor. Consider inspecting: batchSize." is one example.

**Comment by gauravghati** [comment]
> You can click on the static code analysis checks below to see the report. "Not all members of a class are initialized inside the constructor. Consider inspecting: batchSize." is one example.

yes, thanks, solved errors and warnings from static code analysis report, let me know if I left anything.

**Comment by gauravghati** [comment]
@zoq Please review, whenever you have time.

File: src/mlpack/methods/ann/layer/adaptive_max_pooling_impl.hpp [hunk 2]
```diff
@@ -33,6 +36,56 @@ AdaptiveMaxPooling<InputDataType, OutputDataType>::AdaptiveMaxPooling(
...
+    outputWidth(std::move(layer.outputWidth)),
+    outputHeight(std::move(layer.outputHeight)),
+    reset(std::move(layer.reset))
+{
+  // Nothing to do here.
+}
+
+template <typename InputDataType, typename OutputDataType>
+AdaptiveMaxPooling<InputDataType, OutputDataType>& 
+AdaptiveMaxPooling<InputDataType, OutputDataType> :: operator=(const AdaptiveMaxPooling& layer)
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/adaptive_max_pooling_impl.hpp:63]:
```suggestion
AdaptiveMaxPooling<InputDataType, OutputDataType>::operator=(
  const AdaptiveMaxPooling& layer)
```

Remove some extra spaces and align with the 80 character line limit.

File: src/mlpack/methods/ann/layer/adaptive_max_pooling_impl.hpp [hunk 3]
```diff
@@ -33,6 +36,56 @@ AdaptiveMaxPooling<InputDataType, OutputDataType>::AdaptiveMaxPooling(
...
+    outputWidth = layer.outputWidth;
+    outputHeight = layer.outputHeight;
+    reset = layer.reset;
+  }
+  return *this; 
+}
+
+template <typename InputDataType, typename OutputDataType>
+AdaptiveMaxPooling<InputDataType, OutputDataType>& 
+AdaptiveMaxPooling<InputDataType, OutputDataType> :: operator=(AdaptiveMaxPooling&& layer)
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/adaptive_max_pooling_impl.hpp:77]:
```suggestion
AdaptiveMaxPooling<InputDataType, OutputDataType>::operator=(
  AdaptiveMaxPooling&& layer)
```

See comment above.

File: src/mlpack/methods/ann/layer/convolution_impl.hpp [hunk 4]
```diff
@@ -138,6 +138,172 @@ Convolution<
...
+    InputDataType,
+    OutputDataType
+>& 
+Convolution<
+    ForwardConvolutionRule,
+    BackwardConvolutionRule,
+    GradientConvolutionRule,
+    InputDataType,
+    OutputDataType
+> :: operator=(const Convolution& layer)
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/convolution_impl.hpp:235]:
```suggestion
>&
Convolution<
    ForwardConvolutionRule,
    BackwardConvolutionRule,
    GradientConvolutionRule,
    InputDataType,
    OutputDataType
>::operator=(const Convolution& layer)
```

See comment above.

File: src/mlpack/methods/ann/layer/convolution_impl.hpp [hunk 5]
```diff
@@ -138,6 +138,172 @@ Convolution<
...
+    InputDataType,
+    OutputDataType
+>& 
+Convolution<
+    ForwardConvolutionRule,
+    BackwardConvolutionRule,
+    GradientConvolutionRule,
+    InputDataType,
+    OutputDataType
+> :: operator=(Convolution&& layer)
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/convolution_impl.hpp:281]:
```suggestion
> :operator=(Convolution&& layer)
```

See comment above.

File: src/mlpack/methods/ann/layer/convolution_impl.hpp [hunk 6]
```diff
@@ -138,6 +138,172 @@ Convolution<
...
+    InputDataType,
+    OutputDataType
+>& 
+Convolution<
+    ForwardConvolutionRule,
+    BackwardConvolutionRule,
+    GradientConvolutionRule,
+    InputDataType,
+    OutputDataType
+> :operator=(Convolution&& layer)
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/convolution_impl.hpp:281]: Hey @gauravghati, this looks like the line that's causing the problem with the build.  If you can get the build worked out, I'd be glad to help move this PR forward and get it merged. :+1:

**Pull request closed without merging**